MaxText/layers/linears.py

```diff
@@ -352,7 +358,7 @@ def unpermute(self, intermediate, sorted_selected_experts, weights):
     tensor_parallelism = self.config.ici_tensor_parallelism * self.config.dcn_tensor_parallelism
     reshaped_intermediate = jnp.reshape(unsort_intermediate, (-1, self.num_experts_per_tok, self.config.emb_dim // tensor_parallelism))
     with jax.named_scope("weight_sum"):
-      output = jnp.einsum("BKE,BK -> BE", reshaped_intermediate, reshaped_weights)
```

Comment:
this is still an einsum (matmul), so the float32 doesn't take full affect without setting additional flags. I think we could instead broadcast the weights to BKE, elementwise matmul (bf16 is prob fine for elementwise matmul), then sum in float32.

Alternatively since this is a matmul you can pass in the matmul precision arg here, although I think this op should be set to f32 independent of the matmul precision

Reply:
ok, I will directly add precision arg there at this moment then. Thanks Matt!